src/rust/engine/process_execution/src/tests.rs

```diff
@@ -88,3 +88,21 @@ fn process_result_metadata_to_and_from_executed_action_metadata() {
   let process_result_missing: ExecutedActionMetadata = ProcessResultMetadata::default().into();
   assert_eq!(process_result_missing, ExecutedActionMetadata::default());
 }
+
+#[test]
+fn process_result_metadata_time_saved_from_cache() {
+  let metadata = ProcessResultMetadata::new(Some(concrete_time::Duration::new(5, 150)));
+  let time_saved = metadata.time_saved_from_cache(Duration::new(1, 100));
+  assert_eq!(time_saved, Some(Duration::new(4, 50)));
+
+  // If the original process time wasn't recorded, we can't compute the time saved.
+  assert_eq!(
+    ProcessResultMetadata::default().time_saved_from_cache(Duration::new(1, 100)),
+    None
+  );
```

Comment:
Should we consider storing a 0 here instead (similar to the negative case)? The idea would be to always have an entry in this histogram when we end up using the cache, even if we can't report a benefit.

Reply:
I don't think so, wouldn't that skew the data? Let's say we have 50 cache hits, only 30 of which had the original execution time recorded. That means that 60% of our histogram entries will be 0 and will skew the data to look like the caching is not very helpful.

Fwit, local execution should always be storing the original execution time. The only reason we might not have the execution time is that we can't guarantee remote execution has set the `ExecutedActionMetadata`. But `cache.rs` and `remote_cache.rs` only ever deal with local process executions, so in practicality, those codepaths will always have access to the original time.

Reply:
> I don't think so, wouldn't that skew the data? Let's say we have 50 cache hits, only 30 of which had the original execution time recorded. That means that 60% of our histogram entries will be 0 and will skew the data to look like the caching is not very helpful.

It wouldn't skew the data... it would make it more accurate, if anything: if there are a bunch of zeros being recorded, we need to expose that, as it could be a problem in one of a few different places.


>> It's possible to be negative if the current process run takes longer than the previous one did.
> 
> When would that happen?

In exactly the case I've described, I think. You ran the process in the past: it took 1.5 seconds to run. You're running it again, and racing it against a cache lookup. For some reason, when you run the process again, it takes a lot longer to run than it did the first time... say 3 seconds. Meanwhile, your cache lookup also takes a lot longer, but still wins the race in 2 seconds. You end up with a time saved value of `1.5 - 2`.

Reply:
Okay so there are two different scenarios and I think we want to handle them differently:

1. The original time was not recorded. As explained above, this _should_ never happen because local execution should always have the timing recorded, and `cache.rs` and `remote_cache.rs` solely work with local processes, not remote caching. We're only handling this case for type safety - if you want, we could error or use `Result` in this case, so that we eagerly know when the original process time wasn't recorded? I'm not comfortable with using 0 to represent this case.
2. The cache hit was slower than the original process, but still faster than rerunning the process (via speculation). This is the negative numbers case. We, by definition, cannot know how slow the process would be on the second run, so the best we can do is represent with a 0. That sounds sensible to me.

So, we could change the modeling of `.time_saved_from_cache()` be `Result<Duration, String>`, where a Duration of 0 means the cache hit was slower than original process. Or stick with `Option<String>` and use 0. Wdyt?

Reply:
> The original time was not recorded. As explained above, this should never happen because local execution should always have the timing recorded, and cache.rs and remote_cache.rs solely work with local processes, not remote caching. We're only handling the missing data case for type safety - if you want, we could error or use Result in this case, so that we eagerly know when the original process time wasn't recorded? I'm not comfortable with using 0 to represent this case.

It can happen in the context of remote execution, where a server does not return the value. We shouldn't error for a server that cannot return us metrics, but we should do something to indicate that the server is reporting bad metrics.

> So, we could change the modeling of .time_saved_from_cache() be Result<Duration, String>, where a Duration of 0 means the cache hit was slower than original process. Or stick with Option<String> and use 0. Wdyt?

Warnings would be annoying, and we'd need to add a way to disable them probably. Could put it at debug, but then fair chance that we never notice it. So yea, I'd still vote for 0 here, as it exposes the issue iff someone is actually looking at these metrics... but debug level logging wouldn't be the end of the world.

> That sounds sensible to me.

👍

Reply:
> It can happen in the context of remote execution, where a server does not return the value. 

Yes, but `cache.rs` and `remote_cache.rs` never will consume the results of remote execution, right? I'm only talking about case #1 here. 

K, I'll change case #2 to use 0, but keep Option to represent case #1.

Reply:
> Yes, but cache.rs and remote_cache.rs never will consume the results of remote execution, right? I'm only talking about case #1 here.

They will if a local cache is wrapped around remote execution, which is common.

Reply:
> They will if a local cache is wrapped around remote execution, which is common.

I'm not familiar with this workflow - when does this happen?

Reply:
https://github.com/pantsbuild/pants/blob/e19336db0d3df6361e08c53b921da7c07a880045/src/rust/engine/src/context.rs#L157-L218

`maybe_local_cached_command_runner` is wrapped around `maybe_remote_enabled_command_runner`.